L1Trigger/Phase2L1GMT/interface/Isolation.h

```diff
@@ -0,0 +1,245 @@
+    Isolation(const Isolation &cpy);
+
+    unsigned compute_trk_iso(l1t::TrackerMuon &in_mu, ConvertedTTTrack &in_trk);
+
+    void isolation_allmu_alltrk(std::vector<l1t::TrackerMuon> &trkMus, std::vector<ConvertedTTTrack> &convertedTracks);
+
+  private:
+    void DumpOutputs(std::vector<l1t::TrackerMuon> &trkMus);
+    int SetAbsIsolationBits(int accum);
+    int SetRelIsolationBits(int accum, int mupt);
+    int OverlapRemoval(unsigned &ovrl, std::vector<unsigned> &overlaps);
+
+    const static int c_iso_dangle_max = 260;  //@ <  260 x 2pi/2^13 = 0.2 rad
+    const static int c_iso_dz_max = 17;       //@ <  17 x 60/2^10 = 1   cm
+    const static int c_iso_pt_min = 120;      //@ >= , 120  x 25MeV = 3 GeV
+
+    // Assuming 4 bits for Muon isolation
+    int absiso_thrL;
+    int absiso_thrM;
+    int absiso_thrT;
+    double reliso_thrL;
+    double reliso_thrM;
+    double reliso_thrT;
+    bool verbose_;
+    bool dumpForHLS_;
+    std::ofstream dumpOutput;
+
+    typedef ap_ufixed<9, 9, AP_TRN, AP_SAT> iso_accum_t;
+    typedef ap_ufixed<9, 0> reliso_thresh_t;
+  };
+
+  inline Isolation::Isolation(const edm::ParameterSet &iConfig)
+      : absiso_thrL(iConfig.getParameter<int>("AbsIsoThresholdL")),
+        absiso_thrM(iConfig.getParameter<int>("AbsIsoThresholdM")),
+        absiso_thrT(iConfig.getParameter<int>("AbsIsoThresholdT")),
+        reliso_thrL(iConfig.getParameter<double>("RelIsoThresholdL")),
+        reliso_thrM(iConfig.getParameter<double>("RelIsoThresholdM")),
+        reliso_thrT(iConfig.getParameter<double>("RelIsoThresholdT")),
+        verbose_(iConfig.getParameter<int>("verbose")),
+        dumpForHLS_(iConfig.getParameter<int>("IsodumpForHLS")) {
+    dumpForHLS_ = true;
+    if (dumpForHLS_) {
+      dumpInput.open("Isolation_Mu_Track_infolist.txt", std::ofstream::out);
+      dumpOutput.open("Isolation_Mu_Isolation.txt", std::ofstream::out);
+    }
+  }
+
+  inline Isolation::~Isolation() {
+    if (dumpForHLS_) {
+      dumpInput.close();
+      dumpOutput.close();
+    }
+  }
+
+  inline Isolation::Isolation(const Isolation &cpy) : TopoAlgo(cpy) {}
+
+  inline void Isolation::DumpOutputs(std::vector<l1t::TrackerMuon> &trkMus) {
+    static int nevto = 0;
+    for (unsigned int i = 0; i < trkMus.size(); ++i) {
+      auto mu = trkMus.at(i);
+      if (mu.hwPt() != 0) {
+        double convertphi = mu.hwPhi() * LSBphi;
+        if (convertphi > M_PI) {
+          convertphi -= 2 * M_PI;
+        }
+        dumpOutput << nevto << " " << i << " " << mu.hwPt() * LSBpt << " " << mu.hwEta() * LSBeta << " " << convertphi
+                   << " " << mu.hwZ0() * LSBGTz0 << " " << mu.hwIso() << endl;
+      }
+    }
+    nevto++;
+  }
+
+  inline int Isolation::SetAbsIsolationBits(int accum) {
+    int iso = (accum <= absiso_thrT ? 3 : accum <= absiso_thrM ? 2 : accum <= absiso_thrL ? 1 : 0);
+
+    if (verbose_) {
+      cout << " [DEBUG Isolation] : absiso_threshold L : " << absiso_thrL << " accum " << accum
```

Comment:
Here and other places:  std::cout is not threadsafe, please use https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideMessageLogger